File: ios/Gogai/Utilities/AppCache.swift
import Foundation

/// アプリのローカルキャッシュを管理するクラス
/// 記事・フィード・グループをCachesディレクトリにJSONファイルとして保存する
///
/// `@unchecked Sendable` の根拠: 内部状態は初期化後に変更されない `directory: URL` のみ。
/// 将来可変プロパティを追加する場合は `NSLock` 等によるスレッド安全性の確保が必要。
final class AppCache: @unchecked Sendable {
    static let shared = AppCache()

    private let directory: URL

    init(directory: URL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("gogai")) {
        self.directory = directory
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    }

    // MARK: - Articles

    func saveAllArticles(_ articles: [Article]) {
        save(articles, to: "allArticles.json")
    }

    func loadAllArticles() -> [Article] {
        load([Article].self, from: "allArticles.json") ?? []
    }

    // MARK: - FeedCounts

    func saveFeedCounts(_ counts: [FeedCount]) {
        save(counts, to: "feedCounts.json")
    }

    func loadFeedCounts() -> [FeedCount] {
        load([FeedCount].self, from: "feedCounts.json") ?? []
    }

    // MARK: - Feeds

    func saveFeeds(_ feeds: [Feed]) {
        save(feeds, to: "feeds.json")
    }

    func loadFeeds() -> [Feed] {
        load([Feed].self, from: "feeds.json") ?? []
    }

    // MARK: - Groups

    func saveGroups(_ groups: [Group]) {
        save(groups, to: "groups.json")
    }

    func loadGroups() -> [Group] {
        load([Group].self, from: "groups.json") ?? []
    }

    // MARK: - Stocks

    func saveStocks(_ stocks: [Stock]) {
        save(stocks, to: "stocks.json")
    }

    func loadStocks() -> [Stock] {
        load([Stock].self, from: "stocks.json") ?? []
    }

    func saveStockCategories(_ categories: [StockCategory]) {
        save(categories, to: "stockCategories.json")
    }

    func loadStockCategories() -> [StockCategory] {
        load([StockCategory].self, from: "stockCategories.json") ?? []
    }

    // MARK: - Size

    var totalSize: Int64 {
        guard let enumerator = FileManager.default.enumerator(
            at: directory,
            includingPropertiesForKeys: [.fileSizeKey],
            options: .skipsHiddenFiles
        ) else { return 0 }
        return enumerator.compactMap { $0 as? URL }
            .compactMap { try? $0.resourceValues(forKeys: [.fileSizeKey]).fileSize }
            .reduce(0) { $0 + Int64($1) }
    }

    // MARK: - Clear

    func clearAll() {
        try? FileManager.default.removeItem(at: directory)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    }

    // MARK: - Private

    private func save<T: Encodable>(_ value: T, to filename: String) {
        let url = directory.appendingPathComponent(filename)
        try? JSONEncoder().encode(value).write(to: url, options: .atomic)
    }

    private func load<T: Decodable>(_ type: T.Type, from filename: String) -> T? {
        let url = directory.appendingPathComponent(filename)
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(type, from: data)
    }
}

File: android/app/src/main/kotlin/com/mtkg/gogai/cache/AppCache.kt
package com.mtkg.gogai.cache

import com.mtkg.gogai.model.Article
import com.mtkg.gogai.model.Feed
import com.mtkg.gogai.model.FeedCount
import com.mtkg.gogai.model.Group
import com.mtkg.gogai.model.Stock
import com.mtkg.gogai.model.StockCategory
import java.io.File
import kotlinx.serialization.KSerializer
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/// アプリのローカルキャッシュ（iOS AppCache の移植）。
/// 記事・フィード・グループを cacheDir/gogai に JSON ファイルとして保存する。
class AppCache(baseDir: File) {
    private val directory = File(baseDir, "gogai")
    private val json = Json { ignoreUnknownKeys = true }

    init {
        directory.mkdirs()
    }

    fun saveAllArticles(articles: List<Article>) = save(articles, ListSerializer(Article.serializer()), "allArticles.json")
    fun loadAllArticles(): List<Article> = load(ListSerializer(Article.serializer()), "allArticles.json") ?: emptyList()

    fun saveFeedCounts(counts: List<FeedCount>) = save(counts, ListSerializer(FeedCount.serializer()), "feedCounts.json")
    fun loadFeedCounts(): List<FeedCount> = load(ListSerializer(FeedCount.serializer()), "feedCounts.json") ?: emptyList()

    fun saveFeeds(feeds: List<Feed>) = save(feeds, ListSerializer(Feed.serializer()), "feeds.json")
    fun loadFeeds(): List<Feed> = load(ListSerializer(Feed.serializer()), "feeds.json") ?: emptyList()

    fun saveGroups(groups: List<Group>) = save(groups, ListSerializer(Group.serializer()), "groups.json")
    fun loadGroups(): List<Group> = load(ListSerializer(Group.serializer()), "groups.json") ?: emptyList()

    fun saveStocks(stocks: List<Stock>) = save(stocks, ListSerializer(Stock.serializer()), "stocks.json")
    fun loadStocks(): List<Stock> = load(ListSerializer(Stock.serializer()), "stocks.json") ?: emptyList()

    fun saveStockCategories(categories: List<StockCategory>) =
        save(categories, ListSerializer(StockCategory.serializer()), "stockCategories.json")
    fun loadStockCategories(): List<StockCategory> =
        load(ListSerializer(StockCategory.serializer()), "stockCategories.json") ?: emptyList()

    val totalSize: Long
        get() = directory.walkTopDown().filter { it.isFile }.sumOf { it.length() }

    fun clearAll() {
        directory.deleteRecursively()
        directory.mkdirs()
    }

    private fun <T> save(value: T, serializer: KSerializer<T>, filename: String) {
        runCatching {
            // 書き込み途中のクラッシュで壊れたファイルを残さないよう一時ファイル経由で置き換える
            val tmp = File(directory, "$filename.tmp")
            tmp.writeText(json.encodeToString(serializer, value))
            tmp.renameTo(File(directory, filename))
        }
    }

    private fun <T> load(serializer: KSerializer<T>, filename: String): T? {
        val file = File(directory, filename)
        if (!file.exists()) return null
        return runCatching { json.decodeFromString(serializer, file.readText()) }.getOrNull()
    }
}
